# Assignment 2: Linear Models and Validation Metrics
Due Date: October 10, 2023
<br>
<br>
Name: Christian Valdez

## Part 1: Classification (14.5 marks total)

### Step 0: Import Libraries

In [1]:
# common libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# dataset
import yellowbrick
from yellowbrick.datasets import load_spam, load_concrete

# sklearn
from sklearn.model_selection import train_test_split

### Step 1: Data Input (1 mark)

The data used for this task can be downloaded using the yellowbrick library: 
https://www.scikit-yb.org/en/latest/api/datasets/spam.html

In [27]:
# import spam dataset from yellowbrick library
X, y = yellowbrick.datasets.loaders.load_spam(
    data_home=None, return_dataset=False)

In [3]:
# print size of X
rows, cols = X.shape
data_size = X.size
print(f"There are {rows} samples and {cols} features.\nThe size of the feature matrix is {data_size}.")

There are 4600 samples and 57 features.
The size of the feature matrix is 262200.


In [34]:
# print size of y
rows, = y.shape
print(f"The feature matrix comes with the corresponding {rows} labels.\n0 indicating a not spam email (ham) and 1 for a spam email.")

The feature matrix comes with the corresponding 4600 labels.
0 indicating a not spam email (ham) and 1 for a spam email.


In [5]:
# feature data type of X
print("Data type for features: ", )
for feature, dtype in X.dtypes.items():
    print(f"{feature}: {dtype}")

Data type for features: 
word_freq_make: float64
word_freq_address: float64
word_freq_all: float64
word_freq_3d: float64
word_freq_our: float64
word_freq_over: float64
word_freq_remove: float64
word_freq_internet: float64
word_freq_order: float64
word_freq_mail: float64
word_freq_receive: float64
word_freq_will: float64
word_freq_people: float64
word_freq_report: float64
word_freq_addresses: float64
word_freq_free: float64
word_freq_business: float64
word_freq_email: float64
word_freq_you: float64
word_freq_credit: float64
word_freq_your: float64
word_freq_font: float64
word_freq_000: float64
word_freq_money: float64
word_freq_hp: float64
word_freq_hpl: float64
word_freq_george: float64
word_freq_650: float64
word_freq_lab: float64
word_freq_labs: float64
word_freq_telnet: float64
word_freq_857: float64
word_freq_data: float64
word_freq_415: float64
word_freq_85: float64
word_freq_technology: float64
word_freq_1999: float64
word_freq_parts: float64
word_freq_pm: float64
word_freq_direc

In [31]:
# feature data type of y
print("Data type for labels: ", y.dtypes)

Data type for labels:  int64


### Step 2: Data Processing (1.5 marks)

In [32]:
# check for missing values
missing_values_X = X.isnull().sum().sum()
missing_values_y = y.isnull().sum()

if missing_values_X != 0:
    X.fillna(0, inplace=True)
    print(f"Filled missing values in X. Total filled: {missing_values_X}")
else:
    print("There are no missing values in X.")

if missing_values_y != 0:
    y.fillna(0, inplace=True)
    print(f"Filled missing values in y. Total filled: {missing_values_y}")
else:
    print("There are no missing values in y.")

There are no missing values in X.
There are no missing values in y.


In [33]:
# split data with test size 5% of the data
X_train, X_small, y_train, y_small = train_test_split(
    X, y, test_size=0.05, random_state=0)

print(f"Training set size: {len(X_small)}")
print(f"Test set size: {len(y_small)}")

Training set size: 230
Test set size: 230


### Step 3: Implement Machine Learning Model

1. Import `LogisticRegression` from sklearn
2. Instantiate model `LogisticRegression(max_iter=2000)`.
3. Implement the machine learning model with three different datasets: 
    - `X` and `y`
    - Only first two columns of `X` and `y`
    - `X_small` and `y_small`

### Step 4: Validate Model

### Step 5: Visualize Results (4 marks)

1. Create a pandas DataFrame `results` with columns: Data size, training accuracy, validation accuracy
2. Add the data size, training and validation accuracy for each dataset to the `results` DataFrame
3. Print `results`

In [9]:
# import
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# get the first two columns
x_first_two_columns = X.iloc[:, :2]
x_first_two_columns.head()

# results dataframe
results = {
    "Data Size": [],
    "Training Accuracy": [],
    "Validation Accuracy": []
}

# datasets
X_datasets = [X, x_first_two_columns, X_small]
X_labels = [y, y, y_small]
X_dataset_names = ["Full Dataset", "First Two Columns", "5% of the dataset"]

for dataset_num in range(3):
    row, col = X_datasets[dataset_num].shape

    print(f"Training with {X_dataset_names[dataset_num]}")
    x1, x2, y1, y2 = train_test_split(X_datasets[dataset_num], X_labels[dataset_num], test_size=0.2, random_state=0)

    # instantiate the model
    model = LogisticRegression(max_iter=2000)

    # train the dataset
    model.fit(x1, y1)

    # training accuracy
    training_score = model.score(x1, y1)

    # validation accuracy
    validation_score = model.score(x2, y2)

    # appends to results table
    results["Data Size"].append(row*col)
    results["Training Accuracy"].append(training_score)
    results["Validation Accuracy"].append(validation_score)

results_df = pd.DataFrame(results)
results_df.index = X_dataset_names

Training with Full Dataset


Training with First Two Columns
Training with 5% of the dataset


In [10]:
results_df

Data Size  Training Accuracy  Validation Accuracy
Full Dataset          262200           0.927989             0.936957
First Two Columns       9200           0.614946             0.593478
5% of the dataset      13110           0.956522             0.804348

### Questions (4 marks)
1. How do the training and validation accuracy change depending on the amount of data used? Explain with values.

With the full dataset, the training and validation accuracies are quite similar, registering at **0.928** and **0.937** respectively. Interestingly, the training accuracy is slightly lower, possibly suggesting that the training data might be more intricate than the validation set. 

Using only 5% of the data, there's a noticeable divergence between the training and validation accuracies — **0.957** versus **0.804**. This disparity suggests potential overfitting.

Reducing the features yields diminished accuracies for both training and validation, recorded at **0.615** and **0.593**, respectively.

2. In this case, what do a false positive and a false negative represent? Which one is worse?

A false positive indicates that the model erroneously labeled an email as spam when it wasn't. Conversely, a false negative means the model missed identifying a genuine spam email. The latter, false negatives, are more detrimental since overlooking spam emails can introduce security risks. While mislabeling a genuine email as spam isn't ideal, it's less concerning than the alternative.

### Process Description (4 marks)
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code?
1. In what order did you complete the steps?
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?

*DESCRIBE YOUR PROCESS HERE*
1. I used ChatGPT to review my code and help clarify some questions. Also, I used common Python documentation websites to search for built in functions.

- [ChatGPT](https://chat.openai.com/share/06987052-fc84-4545-8c0b-ef26077cca12)

## Part 2: Regression (10.5 marks total)

For this section, we will be evaluating concrete compressive strength of different concrete samples, based on age and ingredients. You will need to repeat the steps 1-4 from Part 1 for this analysis.

### Step 1: Data Input (1 mark)

The data used for this task can be downloaded using the yellowbrick library: 
https://www.scikit-yb.org/en/latest/api/datasets/concrete.html

In [11]:
# import  dataset
X, y = yellowbrick.datasets.loaders.load_concrete(
    data_home=None, return_dataset=False)

In [12]:
# feature matrix
rows, cols = X.shape
data_size = X.size
print(
    f"There are {rows} samples and {cols} features.\nThe size of the feature matrix is {data_size}.")


There are 1030 samples and 8 features.
The size of the feature matrix is 8240.


### Step 2: Data Processing (0.5 marks)

In [13]:
# check for missing values
missing_values_X = X.isnull().sum().sum()
missing_values_y = y.isnull().sum()

if missing_values_X != 0:
    X.fillna(0, inplace=True)
    print(f"Filled missing values in X. Total filled: {missing_values_X}")
else:
    print("There are no missing values in X.")

if missing_values_y != 0:
    y.fillna(0, inplace=True)
    print(f"Filled missing values in y. Total filled: {missing_values_y}")
else:
    print("There are no missing values in y.")


There are no missing values in X.
There are no missing values in y.


### Step 3: Implement Machine Learning Model (1 mark)

In [14]:
#import Linear Regression
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.metrics import mean_squared_error, r2_score

In [15]:
# split data train = 80%, test = 20%
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=0)

In [16]:
# initialize model
linear_model = LinearRegression()

# train model
linear_model.fit(X_train, y_train)

LinearRegression()

### Step 4: Validate Model (1 mark)

Calculate the training and validation accuracy using mean squared error and R2 score.

In [17]:
# make predictions
linear_prediction_train = linear_model.predict(X_train)
linear_prediction_test = linear_model.predict(X_test)

# evaluate the model
mse_train = mean_squared_error(y_train, linear_prediction_train)
mse_test = mean_squared_error(y_test, linear_prediction_test)

# Compute and print the R^2 score
r2_train = r2_score(y_train, linear_prediction_train)
r2_test = r2_score(y_test, linear_prediction_test)

### Step 5: Visualize Results (1 mark)

In [18]:
results_lr = {
    "Training Accuracy": [mse_train, r2_train],
    "Validation Accuracy": [mse_test, r2_test]
}
results_lr = pd.DataFrame(results_lr)
results_lr.index = ["MSE", "R2 Score"]

In [19]:
results_lr

Training Accuracy  Validation Accuracy
MSE              111.358439            95.904136
R2 Score           0.610823             0.623414

### Questions (2 marks)
1. Did using a linear model produce good results for this dataset? Why or why not?

### Process Description (4 marks)
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code?
1. In what order did you complete the steps?
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?

*DESCRIBE YOUR PROCESS HERE*

## Part 3: Observations/Interpretation (3 marks)

Describe any pattern you see in the results. Relate your findings to what we discussed during lectures. Include data to justify your findings.


*ADD YOUR FINDINGS HERE*

## Part 4: Reflection (2 marks)
Include a sentence or two about:
- what you liked or disliked,
- found interesting, confusing, challangeing, motivating
while working on this assignment.


*ADD YOUR THOUGHTS HERE*

## Part 5: Bonus Question (4 marks)

Repeat Part 2 with Ridge and Lasso regression to see if you can improve the accuracy results. Which method and what value of alpha gave you the best R^2 score? Is this score "good enough"? Explain why or why not.

**Remember**: Only test values of alpha from 0.001 to 100 along the logorithmic scale.

In [20]:
# Initialize and train a Ridge regression model
ridge_model = Ridge(alpha=100)  # You can tune the alpha parameter
ridge_model.fit(X_train, y_train)

# make predictions
ridge_prediction_train = ridge_model.predict(X_train)
ridge_prediction_test = ridge_model.predict(X_test)

# evaluate the model
mse_train = mean_squared_error(y_train, ridge_prediction_train)
mse_test = mean_squared_error(y_test, ridge_prediction_test)

# Compute and print the R^2 score
r2_train = r2_score(y_train, ridge_prediction_train)
r2_test = r2_score(y_test, ridge_prediction_test)

In [21]:
results_ridge = {
    "Training Accuracy": [mse_train, r2_train],
    "Validation Accuracy": [mse_test, r2_test]
}
results_ridge = pd.DataFrame(results_ridge)
results_ridge.index = ["MSE", "R2 Score"]

In [22]:
results_ridge

Training Accuracy  Validation Accuracy
MSE              111.358548            95.894268
R2 Score           0.610823             0.623453

In [23]:
# Initialize and train a Lasso regression model
lasso_model = Lasso(alpha=1e-4, max_iter=100)
lasso_model.fit(X_train, y_train)

# make predictions
lasso_prediction_train = lasso_model.predict(X_train)
lasso_prediction_test = lasso_model.predict(X_test)

# evaluate the model
mse_train = mean_squared_error(y_train, lasso_prediction_train)
mse_test = mean_squared_error(y_test, lasso_prediction_test)

# Compute and print the R^2 score
r2_train = r2_score(y_train, lasso_prediction_train)
r2_test = r2_score(y_test, lasso_prediction_test)

c:\Users\chris\anaconda3\envs\ensf-ml\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:628: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 4.283e+04, tolerance: 2.209e+01
  model = cd_fast.enet_coordinate_descent(


In [24]:
results_lasso = {
    "Training Accuracy": [mse_train, r2_train],
    "Validation Accuracy": [mse_test, r2_test]
}
results_lasso = pd.DataFrame(results_lasso)
results_lasso.index = ["MSE", "R2 Score"]

In [25]:
results_lasso

Training Accuracy  Validation Accuracy
MSE              111.358450            95.899893
R2 Score           0.610823             0.623431

*ANSWER HERE*